File: sample_concurrency_book/book.py
import random, sys, time

num_active_clients = 0

class book:
  def __init__(self, bookname):
    self.bookname = bookname

class book_enum:
  books = [book("Gone with the wind"),
           book("How to kill a mock bird"),
           book("Programming Concurrency"),
           book("Operating Systems"),
           book("Introductory Robotics")]
  @staticmethod
  def random(self):
    return random.choice(self.books)

class inventory:
  def __init__(self):
    self.inventory = {}
    
  def print_work(self, seq_num, job_type, job_id, book, old_quan, new_quan):
    if job_type:
      print 'A restocking work( %s , %s -> %s ) is performed by job id = %s and seqNum = %s' %(book.bookname, old_quan, new_quan, job_id, seq_num)
    else:
      print 'A shipping work( %s , %s -> %s ) is performed by job id = %s and seqNum = %s' %(book.bookname, old_quan, new_quan, job_id, seq_num)
    
  def increase(self, book, quantity, seq_num, job_id):
    if book in self.inventory:
      self.inventory[book] = self.inventory.get(book) + quantity
    else:
      self.inventory[book] = quantity
    self.print_work(seq_num, True, job_id, book, self.inventory.get(book) - quantity, self.inventory.get(book))
      
  def decrease(self, book, quantity, seq_num, job_id):
    if book in self.inventory and self.inventory.get(book) >= quantity:
      self.inventory[book] = self.inventory.get(book) - quantity
      self.print_work(seq_num, False, job_id, book, self.inventory.get(book) + quantity, self.inventory.get(book))
      return True
    else:
      return False
    
class work:
  def __init__(self, book, quantity):
    self.book = book
    self.quantity = quantity
    
def runnable_job(job_id, job_type):
  job_length = yield
  works = []
  while job_length > 0:
    work = yield
    works.append(work)
    job_length -= 1
  ser = yield
  inv = yield
  seq_num = yield
  
  while len(works) != 0:
    work = random.choice(works)
    if job_type:
      inv.increase(work.book, work.quantity, seq_num, job_id)
    else:
      while not inv.decrease(work.book, work.quantity, seq_num, job_id):
        yield
    works.remove(work)
    yield
  
  ser.finish(job_type)

class server:
  def __init__(self, max_num_job, inv):
    self.max_num_job = max_num_job
    self.inv = inv
    self.num_restock_job = 0
    self.num_ship_job = 0
    self.runnable_jobs = []
    
  def process(self, runnable_job, job_type):
    if job_type:
      if self.num_restock_job < self.max_num_job:
        runnable_job.send(self.inv)
        runnable_job.send(self.num_restock_job)
        self.num_restock_job += 1
        self.runnable_jobs.append(runnable_job)
        return True
      else:
        return False
    else:
      if self.num_ship_job < self.max_num_job:
        runnable_job.send(self.inv)
        runnable_job.send(self.num_ship_job)
        self.num_ship_job += 1
        self.runnable_jobs.append(runnable_job)
        return True
      else:
        return False
      
  def finish(self, job_type):
    if job_type:
      self.num_restock_job -= 1
    else:
      self.num_ship_job -= 1
      
  def run(self):
    while True:
      if len(self.runnable_jobs) != 0:
        try:
          job = random.choice(self.runnable_jobs)
          job.next()
          yield
        except StopIteration:
          self.runnable_jobs.remove(job)
      else:
        yield

def client(client_id, ser, life_time, max_job_length):
  global num_active_clients
  t0 = time.time()
  result = True
  while (time.time() - t0) * 1000 < life_time: #convert to milliseconds
    if result:
      generated_job = generate_job(client_id, ser, life_time, max_job_length)
      job = generated_job[0]
      job_type = generated_job[1]
    result = ser.process(job, job_type)
    yield
  num_active_clients -= 1
  
def generate_job(client_id, ser, life_time, max_job_length):
  job_id = client_id * (10 ^ (len(str(life_time))))
  job_type = random.choice([True, False])
  job = runnable_job(job_id, job_type)
  job.next()
  job_length = random.randint(1, max_job_length)
  job.send(job_length)
  while job_length > 0:
    book = book_enum.random(book_enum)
    quantity = random.randint(1, 20)
    w = work(book, quantity)
    job.send(w)
    job_length -= 1
    print_work(job_type, book, quantity, job_id)
  job.send(ser)
  return (job, job_type)

def print_work(job_type, book, quantity, job_id):
  if job_type:
    print 'A restocking work( %s , %s ) will be performed by job id = %s' %(book.bookname, quantity, job_id)
  else:
    print 'A shipping work( %s , %s ) will be performed by job id = %s' %(book.bookname, quantity, job_id)

def main():
  usage = """Main [num_of_clients]
    [client_lif_time] 
    [max_num_of_job_for_each_type] 
    [max_job_length]"""
    
  if len(sys.argv) < 4:
    print '%s' %usage
    exit()
    
  global num_active_clients
  num_active_clients = int(sys.argv[1])
  
  inv = inventory()
  ser = server(int(sys.argv[3]), inv)
  
  tasks = [ser.run()]
  for i in range(int(sys.argv[1])):
    cli = client(i, ser, int(sys.argv[2]), int(sys.argv[4]))
    tasks.append(cli)
    
  while num_active_clients > 1:
    try:
      task = random.choice(tasks)
      task.next()
    except StopIteration:
      tasks.remove(task)
  
if __name__ == '__main__':
  main()

File: sample_concurrency_book/Book.scala
package book

import scala.actors._
import Actor._
import scala.util.Random
import scala.collection.mutable.Map

/** A base class for enumeration book objects */
class Book(name:String) {
  def bookname():String = name
}

/** Enumeration book objects **/
case object BOOK1 extends Book("Gone with the wind")
case object BOOK2 extends Book("How to kill a mock bird")
case object BOOK3 extends Book("Programming Concurrency")
case object BOOK4 extends Book("Operating Systems")
case object BOOK5 extends Book("Introductory Robotics")

/** BookEnum object that returns a random book when random() method is called */
object BookEnum {
  def random():Book = {
    ((new Random()).nextInt(5)+1) match {
      case 1 =>
        return BOOK1
      
      case 2 =>
        return BOOK2
        
      case 3 =>
        return BOOK3
        
      case 4 =>
        return BOOK4
        
      case 5 =>
        return BOOK5
    }
  }
}

/** A work class with the book to manipulate and the corresponding quantity */
class Work(book:Book, quantity:Int){
  def getBook():Book = book
  def getQuantity():Int = quantity
}

/** Messages exchanged in the system */
case class restock(book:Book, quantity:Int, runnableJob:Actor)
case class ship(book:Book, quantity:Int, runnableJob:Actor)
case class succeed()
case class finish(jobType:Boolean)
case class issue(runnableJob:RunnableJob, client:Actor)
case class fail()
case class quit()

/** A runnable job class.
 *  A runnable job has a unique jobId that is assigned by the client who creates it
 *  A runnable job has a jobType.
 *    A jobType of "true" indicates all works contained in it are restocking works.
 *    A jobType of "false" indicates all works contained in it are shipping works.
 *  The client that creates a runnable job will assign the server for it to run within
 *  The server for a runnable job to run within will assign the book inventory for it to manipulate
 *  The server for a runnable job to run within will also assign a seqNum (sequential number) to the job
 */
class RunnableJob(jobId:Int, jobType:Boolean) extends Actor {
  var server:Actor = null
  var inventory:Actor = null
  var seqNum:Int = 0
  var works:List[Work] = List()
  var succeedWorks:Int = 0
  
  def addWork(book:Book, quantity:Int) {
    works = works.:+(new Work(book, quantity))
  }
  
  def setServer(server:Actor) {
    this.server = server
  }
  
  def setInventory(inventory:Actor) {
    this.inventory = inventory
  }
  
  def setSeqNum(seqNum:Int) {
    this.seqNum = seqNum
  }
  
  def getJobId():Int = jobId
  
  def getSeqNum():Int = seqNum
  
  def getType():Boolean = jobType
  
  def run() {
    for (work <- works) {
      if (jobType) {
        inventory ! restock(work.getBook, work.getQuantity, this)
      } else {
        inventory ! ship(work.getBook, work.getQuantity, this)
      }
    }
  }
  
  def act () {
    run()
    loop {
      react {
        case succeed() =>
          succeedWorks += 1
          if (works.size == succeedWorks) {
            server ! finish(jobType)
          }
      }
    }
  }
}

/** A book inventory class.
 *  This class uses a map to store the stocking information of all books (BOOK1 - BOOK5)
 *  It supports two different manipulations: increase and decrease
 *    An increase manipulation fulfills a certain restocking work
 *    A decrease manipulation fulfills a certain shipping work if stock is sufficient or delays a certain shipping work if stock is insufficient
 */
class Inventory extends Actor {
  var inventory:Map[Book, Int] = Map()
  
  def printWork(job:RunnableJob, book:Book, oldQuan:Int, newQuan:Int) {
    if (job.getType()) {
      println("A restocking work( " + book.bookname() + " , " + oldQuan + " -> " + newQuan + " ) is performed by job id = " + job.getJobId() + " and seqNum = " + job.getSeqNum())
    } else {
      println("A shipping work( " + book.bookname() + " , " + oldQuan + " -> " + newQuan + " ) is performed by job id = " + job.getJobId() + " and seqNum = " + job.getSeqNum())
    }
  }
  
  def increaseInventory(book:Book, quantity:Int, job:RunnableJob) {
    if (inventory.contains(book)) {
      inventory(book) += quantity
    } else {
      inventory = inventory.+((book, quantity))
    }
    job ! succeed()
    printWork(job, book, inventory(book)-quantity, inventory(book))
  }
  
  def decreaseInventory(book:Book, quantity:Int, job:RunnableJob) {
    if (inventory.contains(book)) {
      if (inventory(book) >= quantity) {
        inventory(book) -= quantity
        job ! succeed()
        printWork(job, book, inventory(book)+quantity, inventory(book))
      } else {
        self ! ship(book, quantity, job)
      }
    } else {
      self ! ship(book, quantity, job)
    }
  }
  
  def act() {
    loop {
      react {
        case restock(book:Book, quantity:Int, runnableJob:Actor) =>
          increaseInventory(book, quantity, runnableJob.asInstanceOf[RunnableJob])
          
        case ship(book:Book, quantity:Int, runnableJob:Actor) =>
          decreaseInventory(book, quantity, runnableJob.asInstanceOf[RunnableJob])
      }
    }
  }
}

/** A server class.
 *  This class processes the jobs posted by clients
 *  It has a maxNumJob which indicates the maximum number of running jobs of each type
 *    e.g. if maxNumJob is 5, at any time, at most 5 shipping jobs and 5 restocking jobs (10 in total) could run in server
 *    When the server reaches maxNumJob for a certain type of job, it declines the requests of processing new job of that type 
 *  It has a book inventory for all jobs posted to manipulate
 */
class Server(maxNumJob:Int, inventory:Actor) extends Actor {
  var numRestockJob:Int = 0
  var numShipJob:Int = 0
  
  def act() {
    loop {
      react {
        case issue(runnableJob:RunnableJob, client:Actor) =>
          if (runnableJob.getType) {
            if (numRestockJob < maxNumJob) {
              runnableJob.setInventory(inventory)
              runnableJob.setServer(this)
              runnableJob.setSeqNum(numRestockJob)
              runnableJob.start()
              client ! succeed()
              numRestockJob += 1
            } else {
              client ! fail()
            }
          } else {
            if (numShipJob < maxNumJob) {
              runnableJob.setInventory(inventory)
              runnableJob.setServer(this)
              runnableJob.setSeqNum(numShipJob)
              runnableJob.start()
              client ! succeed()
              numShipJob += 1
            } else {
              client ! fail()
            }
          }
          
        case finish(jobType:Boolean) =>
          if (jobType) {
            numRestockJob -= 1
          } else {
            numShipJob -= 1
          }          
      }
    }
  }
}

/** A client class
 *  This simulated client class randomly generates some different jobs and requests the server to process them
 *  It has a clientId which is assigned by main function (sequentially from 1 to the number of clients as specified in program argument)
 *  It has a server to make requests to
 *  It has a lifeTime which is specified in program argument as milliseconds
 *    A client only lives for this specific number of time to generate jobs and make requests to server
 *    When this amount of lifeTime expires, a client quite anyway
 *  It has a maxJobLength which is specified in program argument and all jobs it generates contains no more than that amount of works
 */
class Client(clientId:Int, server:Actor, lifeTime:Int, maxJobLength:Int, main:Actor) extends Actor {
  var jobCount = 0
  
  def printWork(jobId:Int, jobType:Boolean, book:Book, quantity:Int) {
    if (jobType) {
      println("A restocking work( " + book.bookname() + " , " + quantity + " ) will be performed by job id = " + jobId)
    } else {
      println("A shipping work( " + book.bookname() + " , " + quantity + " ) will be performed by job id = " + jobId)
    }
  }
  
  def generateJob():RunnableJob = {
    val jobId:Int = clientId*(10^((lifeTime.toString()).length))+jobCount
    val jobType:Boolean = (new Random()).nextBoolean()
    val job:RunnableJob = new RunnableJob(jobId, jobType);
    jobCount += 1
    job.setServer(server);
	var jobLength:Int = (new Random()).nextInt(maxJobLength)+1;
	  while (jobLength > 0) {
	    val book:Book = BookEnum.random()
	    val quantity:Int = (new Random()).nextInt(20) + 1
		job.addWork(book, quantity);
		jobLength -= 1;
		printWork(jobId, jobType, book, quantity)
	  }
	return job;
  }
  
  def checkTime(t0:Long) {
    if (System.currentTimeMillis() - t0 > lifeTime) {
      main ! quit()
      exit()
    }
  }
  
  def act() {
    val t0:Long = System.currentTimeMillis()
    var job:RunnableJob = generateJob()
    server ! issue(job, this)
    loop {
      react {
        case succeed() =>
          job = generateJob()
          server ! issue(job, this)
          checkTime(t0)
          
        case fail() =>
          Thread.sleep(500)
          server ! issue(job, this)
          checkTime(t0)
      }
    }
  }
}

/**
 * The main function of simulation
 */
object Main {
  val usage = """Usage: Main [num_of_client] [client_life_time] [max_num_job]"""
  def main(args:Array[String]) {
    val mainActor:Actor = actor {
	  if (args.length < 4) {
		println("Usage: [num_of_client] [client_life_time] [max_num_running_job_for_each_type] [max_job_length]")
		System.exit(1)
	  }
	  
	  val numOfClient:Int = args(0).toInt
	  val clientLifeTime:Int = args(1).toInt
	  val maxNumJob:Int = args(2).toInt
	  val maxJobLength:Int = args(3).toInt
				
	  val inventory:Inventory = new Inventory()
	  inventory.start()
	  val server:Server = new Server(maxNumJob, inventory)
	  server.start()
		
	  for (i <- 0 to numOfClient toList) {
		val client:Client = new Client(i, server, clientLifeTime, maxJobLength, self)
		client.start()
	  }
		
	  loop {
		react {
		  case quit() =>
		    if (mailboxSize + 1 >= numOfClient) {
		      System.exit(0)
		    } else {
		      self ! quit()
		    }
		}
	  }
    }
  }
}